kernel/src/net/socket/unix/stream/socket.rs

```diff
@@ -187,80 +226,60 @@ impl Socket for UnixStreamSocket {
         let addr = UnixSocketAddr::try_from(socket_addr)?;
 
-        match &mut *self.state.write() {
+        match self.state.write().as_mut() {
             State::Init(init) => init.bind(addr),
-            _ => return_errno_with_message!(
-                Errno::EINVAL,
-                "cannot bind a listening or connected socket"
-            ),
-            // FIXME: Maybe binding a connected socket should also be allowed?
+            State::Connected(connected) => connected.bind(addr),
+            State::Listen(_) => {
+                // Listening sockets are always already bound.
+                addr.already_bound()
+            }
         }
     }
 
     fn connect(&self, socket_addr: SocketAddr) -> Result<()> {
         let remote_addr = UnixSocketAddr::try_from(socket_addr)?.connect()?;
+        let backlog = get_backlog(&remote_addr)?;
 
-        // Note that the Linux kernel implementation locks the remote socket and checks to see if
-        // it is listening first. This is different from our implementation, which locks the local
-        // socket and checks the state of the local socket first.
-        //
-        // The difference may result in different error codes, but it's doubtful that this will
-        // ever lead to real problems.
-        //
-        // See also <https://elixir.bootlin.com/linux/v6.10.4/source/net/unix/af_unix.c#L1527>.
-
-        let client_addr = match &*self.state.read() {
-            State::Init(init) => init.addr().cloned(),
-            State::Listen(_) => {
-                return_errno_with_message!(Errno::EINVAL, "the socket is listening")
-            }
-            State::Connected(_) => {
-                return_errno_with_message!(Errno::EISCONN, "the socket is connected")
-            }
-        };
-
-        // We use the `push_incoming` directly to avoid holding the read lock of `self.state`
-        // because it might call `Thread::yield_now` to wait for connection.
-        loop {
-            let res = push_incoming(&remote_addr, client_addr.clone());
-            match res {
-                Ok(connected) => {
-                    *self.state.write() = State::Connected(connected);
-                    return Ok(());
-                }
-                Err(err) if err.error() == Errno::EAGAIN => {
-                    // FIXME: Calling `Thread::yield_now` can cause the thread to run when the backlog is full,
-                    // which wastes a lot of CPU time. Using `WaitQueue` maybe a better solution.
-                    Thread::yield_now()
-                }
-                Err(err) => return Err(err),
-            }
+        if self.is_nonblocking() {
+            self.try_connect(&backlog)
+        } else {
+            backlog.pause_until(|| self.try_connect(&backlog))
         }
     }
 
     fn listen(&self, backlog: usize) -> Result<()> {
+        const SOMAXCONN: usize = 4096;
+        let backlog = backlog.saturating_add(1).min(SOMAXCONN);
```

Comment:
Why we need to add 1 here?

Reply:
Not sure, but I think it's the Linux behavior. We need this to pass the regression test (i.e. at most `backlog + 1` sockets can be in the backlog).